src/psyclone/psyir/nodes/array_mixin.py

```diff
@@ -155,149 +155,126 @@ def _is_bound_op(self, expr, bound_operator, index):
+
+        For example, if a Fortran array A was declared as A(n) then
+        the stopping value is n and A(:UBOUND(A,1)), A(:n) or A(n)
+        would access that value. The starting value is 1 and
+        A(LBOUND(A,1):), A(1:) or A(1) would access that value.
 
         :param int index: the array index to check.
+        :param str bound_type: the type of bound ("lower" or "upper")
 
-        :returns: True if the array index is a range with its stop \
-            value being UBOUND(array,index) or is an expression that \
-            matches the stop value of the declaration and False \
-            otherwise.
+        :returns: True if the array index access includes the \
+            lower/upper bound of the array declaration and False if it \
+            does not or if it can't be determined.
         :rtype: bool
 
         '''
         self._validate_index(index)
 
-        array_dimension = self.indices[index]
-
-        if not isinstance(array_dimension, Range):
-            # Check for a single element array
-            try:
-                symbol = self.scope.symbol_table.lookup(self.name)
-            except SymbolError:
-                # This exception should not happen but catching it is
-                # useful for certain tests where the PSyIR is not
-                # fully set up.
-                return False
-            shape = symbol.datatype.shape
-            sym_maths = SymbolicMaths.get()
-            if (sym_maths.equal(shape[index].lower, shape[index].upper) and
-                    sym_maths.equal(shape[index].lower, array_dimension)):
+        access_shape = self.indices[index]
+
+        # Is this array access in the form of {UL}BOUND(array, index)?
+        if isinstance(access_shape, Range):
+            if bound_type == "upper":
+                operator = BinaryOperation.Operator.UBOUND
+                access_bound = access_shape.stop
+            else:
+                operator = BinaryOperation.Operator.LBOUND
+                access_bound = access_shape.start
+            if self._is_bound_op(access_bound, operator, index):
                 return True
+
+        # Try to compare the upper/lower bound of the array access
+        # with the upper/lower bound of the array declaration.
+
+        # Finding the array declaration is only supported for an
+        # ArrayReference at the moment.
+        # Import here to avoid circular dependence.
+        # pylint: disable=import-outside-toplevel
+        from psyclone.psyir.nodes.array_reference import ArrayReference
+        if not isinstance(self, ArrayReference):
+            return False
+        # pylint: enable=import-outside-toplevel
+
+        symbol = self.symbol
+        if not isinstance(symbol, DataSymbol):
+            # There is no type information for this symbol
+            # (probably because it originates from a wildcard import).
             return False
 
-        upper = array_dimension.stop
+        datatype = symbol.datatype
```

Comment:
No, because `self.datatype` provides the type of the reference which is not what we want. We want the type of the declaration. We can't remove the earlier lines as they check whether we know the type of the declaration, before attempting to get it.